File: android/src/main/java/com/couchbase/cblitereact/Args/DatabaseResource.java
package com.couchbase.cblitereact.Args;

import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.Database;
import com.couchbase.lite.DatabaseConfiguration;
import com.couchbase.lite.ListenerToken;
import com.couchbase.lite.Query;
import com.couchbase.lite.Replicator;
import com.couchbase.lite.ReplicatorConfiguration;

import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;

public class DatabaseResource {
    private Database database;
    private DatabaseConfiguration configuration;


    private ListenerToken listenerToken;

    //replication
    private HashMap<String,ReplicatorResource> replicators =new HashMap<>();

    //Querys
    private HashMap<Integer,QueryResource> querys =new HashMap<>();

    public DatabaseResource(Database db, DatabaseConfiguration config) {
        this.database = db;
        this.configuration = config;
    }

    public DatabaseResource(Database db) {
        this.database = db;
    }

    public Database getDatabase() {
        return database;
    }

    public void setDatabase(Database database) {
        this.database = database;
    }

    public DatabaseConfiguration getConfiguration() {
        return configuration;
    }

    public void setConfiguration(DatabaseConfiguration configuration) {
        this.configuration = configuration;
    }

    public ListenerToken getListenerToken() {
        return listenerToken;
    }

    public void setListenerToken(ListenerToken listenerToken) {
        this.listenerToken = listenerToken;
    }



    public Replicator getReplicator(String replicatorID) {
        return replicators.get(replicatorID).getReplicator();
    }

    String generateReplicatorConfigHash(Replicator replicator) throws NoSuchAlgorithmException {

        ReplicatorConfiguration replicatorConfiguration = replicator.getConfig();
        StringBuilder sbhash = new StringBuilder();
        sbhash.append(replicatorConfiguration.getDatabase().getName());
        sbhash.append(replicatorConfiguration.isContinuous());
        sbhash.append(replicatorConfiguration.getType());
        sbhash.append(replicatorConfiguration.getChannels());
        sbhash.append(replicatorConfiguration.getDocumentIDs());

        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] messageDigest = md.digest(sbhash.toString().getBytes());
        BigInteger number = new BigInteger(1, messageDigest);
        String hashtext = number.toString(16);

        return hashtext;
    }

    public String setReplicator(Replicator replicator) throws NoSuchAlgorithmException {
        ReplicatorResource setreplicator = new ReplicatorResource();
        setreplicator.setReplicator(replicator);

        String hash = generateReplicatorConfigHash(replicator);
        if(!this.replicators.containsKey(hash))
        this.replicators.put(hash,setreplicator);

        return hash;
    }

    public void removeReplicator(String replicatorId) {
        this.replicators.remove(replicatorId);
    }

    public ListenerToken getReplicatorChangeListenerToken(String replicatorID) {
        return this.replicators.get(replicatorID).getReplicatorChangeListenerToken();
    }

    public void setReplicatorChangeListenerToken(ListenerToken replicatorChangeListenerToken,String replicatorID) {
        this.replicators.get(replicatorID).setReplicatorChangeListenerToken(replicatorChangeListenerToken);
    }

    public String getReplicatorChangeListenerJSFunction(String replicatorID) {
        return this.replicators.get(replicatorID).getReplicatorChangeListenerJSFunction();
    }

    public void setReplicatorChangeListenerJSFunction(String replicatorChangeListenerJSFunction,String replicatorID) {
        this.replicators.get(replicatorID).setReplicatorChangeListenerJSFunction(replicatorChangeListenerJSFunction);
    }

    public Query getQuery(Integer queryID) {
        if(this.querys.get(queryID)!=null)
        return this.querys.get(queryID).getQuery();
        else
        return null;
    }

    public Integer setQuery(Query query) throws CouchbaseLiteException {
        QueryResource queryResource = new QueryResource();
        queryResource.setQuery(query);

        Integer hash = query.explain().hashCode();

        if(!querys.containsKey(hash))
        this.querys.put(hash,queryResource);

        return hash;
    }

    public void removeQuery(Integer queryId) {
        this.querys.remove(queryId);
    }

    public ListenerToken getQueryChangeListenerToken(Integer queryId) {
        return querys.get(queryId).getQueryChangeListenerToken();
    }

    public void setQueryChangeListenerToken(ListenerToken QueryChangeListenerToken,Integer queryID) {
        this.querys.get(queryID).setQueryChangeListenerToken(QueryChangeListenerToken);
    }

    public String getQueryChangeListenerJSFunction(Integer QueryID) {
        return this.querys.get(QueryID).getQueryChangeListenerJSFunction();
    }

    public void setQueryChangeListenerJSFunction(String QueryChangeListenerJSFunction,Integer QueryID) {
        this.querys.get(QueryID).setQueryChangeListenerJSFunction(QueryChangeListenerJSFunction);
    }


}

File: ios/Args/DatabaseResource.swift
//
//  DatabaseResource.swift
//  RNReactNativeCblite
//
//  Created by Umer on 05/12/2021.
//  Copyright © 2021 Facebook. All rights reserved.
//

//import UIKit
import CouchbaseLiteSwift


class DatabaseResource: NSObject {
    private var _database: Database?
    var database: Database? {
        get { return _database }
        set { _database = newValue }
    }
    private var _configuration: DatabaseConfiguration?
    var configuration: DatabaseConfiguration? {
        get { return _configuration }
        set { _configuration = newValue }
    }
    private var _listenerToken: ListenerToken?
    var listenerToken: ListenerToken? {
        get { return _listenerToken }
        set { _listenerToken = newValue }
    }
    private var _replicators:[String:ReplicatorResource]?
    var replicators:[String:ReplicatorResource] = [:]
    
    private var _querys:[Int:QueryResource]?
    var querys : [Int:QueryResource] = [:]

    
    override init() {
        super.init()
    }
    
    init(db: Database, config: DatabaseConfiguration) {
        super.init()
        self.database = db
        self.configuration = config
    }
    
    init(db: Database) {
        super.init()
        self.database = db
    }
    
    func getReplicator(replicatorID: String) -> Replicator? {
        return self.replicators[replicatorID]?.replicator
    }
    
    func setReplicator(replicator: Replicator) -> String? {
        let setReplicator = ReplicatorResource()
        setReplicator.replicator = replicator
        let hash = generateReplicatorConfigHash(replicator: replicator)
        if hash != nil {
            self.replicators.updateValue(setReplicator, forKey: hash!)
        }
        return hash
    }
    
    func generateReplicatorConfigHash(replicator: Replicator) -> String? {
        let replicatorConfiguration = replicator.config
        var sbhash = ""
        sbhash.append(replicatorConfiguration.database.name)
        sbhash.append("\(replicatorConfiguration.continuous)")
        sbhash.append("\(replicatorConfiguration.replicatorType)")
        replicatorConfiguration.channels?.forEach({ channel in
            sbhash.append(channel)
        })
        replicatorConfiguration.documentIDs?.forEach({ id in
            sbhash.append(id)
        })
        let md = sbhash.utf8.md5.rawValue
        return md
    }
    
    func removeReplicator(replicatorId: String) {
        self.replicators.removeValue(forKey: replicatorId)
    }
    
    func setReplicatorChangeListenerToken(replicatorId: String, replicatorChangeListenerToken: ListenerToken?) {
        if self.replicators[replicatorId] != nil {
            self.replicators[replicatorId]!.replicatorChangeListenerToken = replicatorChangeListenerToken
        }
    }
    
    func getReplicatorChangeListenerToken(replicatorId: String) -> ListenerToken? {
        if  self.replicators[replicatorId]  != nil &&  self.replicators[replicatorId]!.replicatorChangeListenerToken != nil {
            return self.replicators[replicatorId]!.replicatorChangeListenerToken!
        }
        return nil
    }
    
    func setReplicatorChangeListenerToken(replicatorId: String, replicatorChangeListenerJSFunction: String) {
        // check for empty string
        
        if self.replicators[replicatorId] != nil  {
            self.replicators[replicatorId]!.replicatorChangeListenerJSFunction = replicatorChangeListenerJSFunction
        }
    }
    
    func getReplicatorChangeListenerJSFunction(replicatorId: String) -> String? {
        if  self.replicators[replicatorId] != nil && self.replicators[replicatorId]!.replicatorChangeListenerToken != nil  {
            return self.replicators[replicatorId]!.replicatorChangeListenerJSFunction!
        }
        return nil
    }
    
    func getQuery(queryID: Int) -> Query? {
        if ((!self.querys.isEmpty)&&(self.querys[queryID] != nil)&&(self.querys[queryID]?.query != nil)){
            return self.querys[queryID]?.query
        }
           return nil
    }
    
    func setQuery(query: Query) throws -> Int {
        let queryResource = QueryResource()
        queryResource.query = query
        var hash = 0
        do {
            hash = try query.explain().hashValue
        
            if (!self.querys.keys.contains(hash)) {
                self.querys.updateValue(queryResource, forKey: hash)
            }
        }
        catch let error {
            throw error
        }
        return hash
    }
    
    func removeQuery(queryId: Int) {
        self.querys.removeValue(forKey: queryId)
    }
    
    func getQueryChangeListenerToken(queryId: Int) -> ListenerToken? {
        return self.querys[queryId]?.queryChangeListenerToken
    }
    
    func setQueryChangeListenerToken(queryChangeListenerToken: ListenerToken?, queryID: Int) {
        self.querys[queryID]?.queryChangeListenerToken = queryChangeListenerToken
    }
    
    func getQueryChangeListenerJSFunction(queryID: Int) -> String? {
        return self.querys[queryID]?.queryChangeListenerJSFunction
    }
    
    func setQueryChangeListenerJSFunction(queryChangeListenerJSFunction: String, queryID: Int) {
        self.querys[queryID]?.queryChangeListenerJSFunction = queryChangeListenerJSFunction
    }
}
